### DISCUSSION TITLE
Multiple Activity Incomes (Improved Join)

### DISCUSSION THREAD

**Discussion by FlorianZick** [elsa-workflows/elsa-core#1521] [Design & tech]
Hey,
is it possible to create a "JOIN" activity, which doesn't just wait for all.
Instead, it should have customizeable Income branchs (like the outcomes branches in the Fork activity). This activity should wait for all and can then identify, which input comes from which incomes branch.

In addition it would be great to have a Fork activity, which can give each branch a different input.

Are these two things possible? And if not, are they planed?

Greetings,

Florian

**Comment by sfmskywalker** [comment]
Hi Florian,

This new Join activity that you described, can you tell a little bit more about why it would be useful for it to have a specific set of income branches? Perhaps a concrete use case might help make me understand.

In any case, activities in Elsa do not have any notion of inbound ports, only outbound ports. However, as I'm designing a new architecture for Elsa 3, it would be interesting to understand how inbound ports might be useful.

Regarding your question about the Fork activity where you can provide a different input per branch, that would be possible to implement as a new activity today. It would probably look similar to the Switch activity where you'd have a list of branches and next to each branch an expression editor to specify the output for each (which would be provided as input into the first connected activity to each branch).

This is not planned, but feel free to create an issue for it. And if you create a PR I'd happily accept it 😃

**Reply by FlorianZick** [reply]
The described new Join activity would be really useful for a big project, I would like to do with Elsa. There, I will have a architecture that looks something like this: A list of strings comes as the Input of a Fork activity. The Fork activity splits the workflow and gives each Output branch a specific string of the list. At each branch, there will be a subworkflow or a few activities that do some calculation based on the given string. As a result, there will be again a string for each subworkflow. At the new Join activity, these three strings should be put back into a list. But at this list, its important, that the strings are in a specific order. So the Join activity should have ports like pos1, pos2, etc. 
I could think of a JavaScript expression, where you can say what to do with the inputs. Or a ImprovedJoinBase class, which you can inherent from, create an own Join class and implement the logic (in my case, creating a list with the inputs) in the onExecute Method.

I think, there are many other use cases out there, where you have a fork activity, that starts several subworkflows. The subworkflows produce a result and you need handle these results in the parent workflow. Of course you could always set the result of a subworkflow as a workflow variable with a new activity, but this can blow up the workflow. Or maybe setting a variable is exactly what you want and with a new Join activity, you can set these variables right there and don't have to use a new SetVariable activity for each subworkflow.

 And that's why I think, this improved Join activity could be a quite useful. I could even think of input ports at many other custom activities, where you join with "WaitForAny" a couple of branches and have like a switch function in the onExecute method and define custom actions, based on the input port. 



Thank you for the explanation, how I could do the Fork activity. In a few months, when the planing phase of the project is done, I can implement one and create a PR :)

I hope I could describe, why I think that Input ports could be good. If not, feel free to ask again :)

**Reply by sfmskywalker** [reply]
Very well described, thank you! I agree, having input ports makes sense. I can't promise it will be added to Elsa 2, but will certainly consider this for Elsa 3 :)

**Reply by FlorianZick** [reply]
Thank you, I am happy to hear that :)
